File: minigit-5/ts/src/diff-algo.ts
// Myers diff algorithm implementation

import { DiffHunk, FileDiff } from './types';

interface EditScript {
  op: 'equal' | 'insert' | 'delete';
  oldLine?: number;
  newLine?: number;
  text: string;
}

// Myers diff algorithm - finds shortest edit script
export function myersDiff(oldLines: string[], newLines: string[]): EditScript[] {
  const n = oldLines.length;
  const m = newLines.length;
  const max = n + m;

  if (max === 0) {
    return [];
  }

  // V[k] = x value of furthest reaching point in diagonal k
  const v: Map<number, number> = new Map();
  v.set(1, 0);

  // Trace stores the V arrays for each depth
  const trace: Map<number, number>[] = [];

  // Forward pass - find shortest edit path
  outer: for (let d = 0; d <= max; d++) {
    trace.push(new Map(v));

    for (let k = -d; k <= d; k += 2) {
      // Choose whether to go down or right
      let x: number;
      if (k === -d || (k !== d && (v.get(k - 1) ?? 0) < (v.get(k + 1) ?? 0))) {
        x = v.get(k + 1) ?? 0;
      } else {
        x = (v.get(k - 1) ?? 0) + 1;
      }

      let y = x - k;

      // Follow diagonal (matching lines)
      while (x < n && y < m && oldLines[x] === newLines[y]) {
        x++;
        y++;
      }

      v.set(k, x);

      if (x >= n && y >= m) {
        break outer;
      }
    }
  }

  // Backtrack to find edit script
  const edits: EditScript[] = [];
  let x = n;
  let y = m;

  // trace[d] contains state BEFORE processing depth d
  // So trace[d] = state after processing depth d-1
  // When backtracking from depth d, we need state after d-1, which is trace[d]
  for (let d = trace.length - 1; d > 0; d--) {
    const vPrev = trace[d]; // State after processing d-1
    const k = x - y;

    let prevK: number;
    if (k === -d || (k !== d && (vPrev.get(k - 1) ?? 0) < (vPrev.get(k + 1) ?? 0))) {
      prevK = k + 1;
    } else {
      prevK = k - 1;
    }

    const prevX = vPrev.get(prevK) ?? 0;
    const prevY = prevX - prevK;

    // Add diagonal moves (equal lines)
    while (x > prevX && y > prevY) {
      x--;
      y--;
      edits.unshift({ op: 'equal', oldLine: x, newLine: y, text: oldLines[x] });
    }

    // Add insert or delete
    if (x === prevX) {
      // Insert
      y--;
      edits.unshift({ op: 'insert', newLine: y, text: newLines[y] });
    } else {
      // Delete
      x--;
      edits.unshift({ op: 'delete', oldLine: x, text: oldLines[x] });
    }
  }

  // Handle d=0: any remaining diagonal matches from the start
  while (x > 0 && y > 0 && oldLines[x - 1] === newLines[y - 1]) {
    x--;
    y--;
    edits.unshift({ op: 'equal', oldLine: x, newLine: y, text: oldLines[x] });
  }

  return edits;
}

// Convert edit script to unified diff hunks
export function editScriptToHunks(edits: EditScript[], contextLines: number = 3): DiffHunk[] {
  if (edits.length === 0) {
    return [];
  }

  const hunks: DiffHunk[] = [];
  let currentHunk: DiffHunk | null = null;
  let lastChangeIdx = -1;

  for (let i = 0; i < edits.length; i++) {
    const edit = edits[i];
    const isChange = edit.op !== 'equal';

    if (isChange) {
      // Start new hunk if needed
      if (!currentHunk || i - lastChangeIdx > contextLines * 2) {
        // Finalize previous hunk with trailing context
        if (currentHunk) {
          // Add trailing context from previous hunk
          const trailingStart = lastChangeIdx + 1;
          const trailingEnd = Math.min(trailingStart + contextLines, i);
          for (let j = trailingStart; j < trailingEnd; j++) {
            if (edits[j].op === 'equal') {
              currentHunk.lines.push(' ' + edits[j].text);
              currentHunk.oldCount++;
              currentHunk.newCount++;
            }
          }
          hunks.push(currentHunk);
        }

        // Start new hunk with leading context
        const contextStart = Math.max(0, i - contextLines);
        let oldStart = 1;
        let newStart = 1;

        // Calculate start positions
        for (let j = 0; j < contextStart; j++) {
          if (edits[j].op === 'delete' || edits[j].op === 'equal') oldStart++;
          if (edits[j].op === 'insert' || edits[j].op === 'equal') newStart++;
        }

        currentHunk = {
          oldStart,
          oldCount: 0,
          newStart,
          newCount: 0,
          lines: [],
        };

        // Add leading context
        for (let j = contextStart; j < i; j++) {
          if (edits[j].op === 'equal') {
            currentHunk.lines.push(' ' + edits[j].text);
            currentHunk.oldCount++;
            currentHunk.newCount++;
          }
        }
      } else {
        // Add intervening context lines to current hunk
        for (let j = lastChangeIdx + 1; j < i; j++) {
          if (edits[j].op === 'equal' && currentHunk) {
            currentHunk.lines.push(' ' + edits[j].text);
            currentHunk.oldCount++;
            currentHunk.newCount++;
          }
        }
      }

      // Add the change
      if (currentHunk) {
        if (edit.op === 'delete') {
          currentHunk.lines.push('-' + edit.text);
          currentHunk.oldCount++;
        } else if (edit.op === 'insert') {
          currentHunk.lines.push('+' + edit.text);
          currentHunk.newCount++;
        }
      }

      lastChangeIdx = i;
    }
  }

  // Finalize last hunk
  if (currentHunk) {
    const trailingStart = lastChangeIdx + 1;
    const trailingEnd = Math.min(trailingStart + contextLines, edits.length);
    for (let j = trailingStart; j < trailingEnd; j++) {
      if (edits[j].op === 'equal') {
        currentHunk.lines.push(' ' + edits[j].text);
        currentHunk.oldCount++;
        currentHunk.newCount++;
      }
    }
    hunks.push(currentHunk);
  }

  return hunks;
}

// Generate unified diff between two strings
export function generateDiff(
  oldContent: string,
  newContent: string,
  oldPath: string,
  newPath: string,
  contextLines: number = 3
): FileDiff {
  const oldLines = oldContent ? oldContent.split('\n') : [];
  const newLines = newContent ? newContent.split('\n') : [];

  // Remove trailing empty line if content ends with newline
  if (oldLines.length > 0 && oldLines[oldLines.length - 1] === '') {
    oldLines.pop();
  }
  if (newLines.length > 0 && newLines[newLines.length - 1] === '') {
    newLines.pop();
  }

  const edits = myersDiff(oldLines, newLines);
  const hunks = editScriptToHunks(edits, contextLines);

  return {
    oldPath,
    newPath,
    hunks,
  };
}

// Format a FileDiff to unified diff string
export function formatDiff(diff: FileDiff): string {
  if (diff.isBinary) {
    return `Binary files ${diff.oldPath} and ${diff.newPath} differ\n`;
  }

  if (diff.hunks.length === 0) {
    return '';
  }

  const lines: string[] = [];

  // Header
  lines.push(`--- a/${diff.oldPath}`);
  lines.push(`+++ b/${diff.newPath}`);

  // Hunks
  for (const hunk of diff.hunks) {
    const oldRange =
      hunk.oldCount === 1 ? `${hunk.oldStart}` : `${hunk.oldStart},${hunk.oldCount}`;
    const newRange =
      hunk.newCount === 1 ? `${hunk.newStart}` : `${hunk.newStart},${hunk.newCount}`;
    lines.push(`@@ -${oldRange} +${newRange} @@`);
    lines.push(...hunk.lines);
  }

  return lines.join('\n') + '\n';
}

// Format diff for new file
export function formatNewFileDiff(content: string, path: string): string {
  const lines: string[] = [];

  lines.push(`--- /dev/null`);
  lines.push(`+++ b/${path}`);

  const contentLines = content.split('\n');
  if (contentLines.length > 0 && contentLines[contentLines.length - 1] === '') {
    contentLines.pop();
  }

  if (contentLines.length > 0) {
    lines.push(`@@ -0,0 +1,${contentLines.length} @@`);
    for (const line of contentLines) {
      lines.push('+' + line);
    }
  }

  return lines.join('\n') + '\n';
}

// Format diff for deleted file
export function formatDeletedFileDiff(content: string, path: string): string {
  const lines: string[] = [];

  lines.push(`--- a/${path}`);
  lines.push(`+++ /dev/null`);

  const contentLines = content.split('\n');
  if (contentLines.length > 0 && contentLines[contentLines.length - 1] === '') {
    contentLines.pop();
  }

  if (contentLines.length > 0) {
    lines.push(`@@ -1,${contentLines.length} +0,0 @@`);
    for (const line of contentLines) {
      lines.push('-' + line);
    }
  }

  return lines.join('\n') + '\n';
}

// Calculate diff stat
export function diffStat(diffs: FileDiff[]): { insertions: number; deletions: number } {
  let insertions = 0;
  let deletions = 0;

  for (const diff of diffs) {
    for (const hunk of diff.hunks) {
      for (const line of hunk.lines) {
        if (line.startsWith('+')) insertions++;
        if (line.startsWith('-')) deletions++;
      }
    }
  }

  return { insertions, deletions };
}

// Format diff stat line
export function formatDiffStat(diffs: FileDiff[]): string {
  const lines: string[] = [];
  let totalInsertions = 0;
  let totalDeletions = 0;

  for (const diff of diffs) {
    let insertions = 0;
    let deletions = 0;

    for (const hunk of diff.hunks) {
      for (const line of hunk.lines) {
        if (line.startsWith('+')) insertions++;
        if (line.startsWith('-')) deletions++;
      }
    }

    totalInsertions += insertions;
    totalDeletions += deletions;

    const path = diff.newPath || diff.oldPath;
    const plusses = '+'.repeat(Math.min(insertions, 50));
    const minuses = '-'.repeat(Math.min(deletions, 50));
    lines.push(` ${path} | ${insertions + deletions} ${plusses}${minuses}`);
  }

  const summary =
    ` ${diffs.length} file${diffs.length !== 1 ? 's' : ''} changed` +
    (totalInsertions > 0 ? `, ${totalInsertions} insertion${totalInsertions !== 1 ? 's' : ''}(+)` : '') +
    (totalDeletions > 0 ? `, ${totalDeletions} deletion${totalDeletions !== 1 ? 's' : ''}(-)` : '');

  return lines.join('\n') + '\n' + summary + '\n';
}

File: minigit-5/py/diff_algo.py
"""Myers diff algorithm implementation."""

from typing import List, Tuple, Optional
from dataclasses import dataclass


@dataclass
class DiffHunk:
    """Represents a hunk in unified diff format."""
    old_start: int
    old_count: int
    new_start: int
    new_count: int
    lines: List[str]  # Lines with +/- prefix


def myers_diff(a: List[str], b: List[str]) -> List[Tuple[str, str]]:
    """
    Compute the diff between two sequences using Myers algorithm.
    Returns list of (op, line) where op is ' ', '+', or '-'.
    """
    n, m = len(a), len(b)

    # Handle edge cases
    if n == 0 and m == 0:
        return []
    if n == 0:
        return [('+', line) for line in b]
    if m == 0:
        return [('-', line) for line in a]

    # Use simple LCS-based approach for smaller inputs
    # Myers is overkill and complex for our needs
    return lcs_diff(a, b)


def lcs_diff(a: List[str], b: List[str]) -> List[Tuple[str, str]]:
    """Compute diff using LCS (simpler than Myers but correct)."""
    n, m = len(a), len(b)

    # Build LCS table
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    # Backtrack to get diff
    result = []
    i, j = n, m
    while i > 0 or j > 0:
        if i > 0 and j > 0 and a[i - 1] == b[j - 1]:
            result.append((' ', a[i - 1]))
            i -= 1
            j -= 1
        elif j > 0 and (i == 0 or dp[i][j - 1] >= dp[i - 1][j]):
            result.append(('+', b[j - 1]))
            j -= 1
        else:
            result.append(('-', a[i - 1]))
            i -= 1

    return list(reversed(result))


def create_unified_diff(
    old_lines: List[str],
    new_lines: List[str],
    old_name: str,
    new_name: str,
    context: int = 3
) -> str:
    """Create unified diff format output."""
    diff_ops = lcs_diff(old_lines, new_lines)

    if not diff_ops:
        return ""

    # Check if there are any actual changes
    has_changes = any(op != ' ' for op, _ in diff_ops)
    if not has_changes:
        return ""

    lines = []
    lines.append(f"--- {old_name}")
    lines.append(f"+++ {new_name}")

    # Group into hunks
    hunks = create_hunks(diff_ops, context)

    for hunk in hunks:
        lines.append(f"@@ -{hunk.old_start},{hunk.old_count} +{hunk.new_start},{hunk.new_count} @@")
        lines.extend(hunk.lines)

    return '\n'.join(lines) + '\n'


def create_hunks(diff_ops: List[Tuple[str, str]], context: int = 3) -> List[DiffHunk]:
    """Group diff operations into hunks with context."""
    if not diff_ops:
        return []

    hunks = []
    current_hunk = None
    old_line = 1
    new_line = 1

    # Find change positions
    change_positions = [i for i, (op, _) in enumerate(diff_ops) if op != ' ']

    if not change_positions:
        return []

    # Process each change, grouping nearby changes into hunks
    i = 0
    while i < len(diff_ops):
        op, line = diff_ops[i]

        if op != ' ':
            # Start or extend a hunk
            if current_hunk is None:
                # Start new hunk with context before
                start_ctx = max(0, i - context)
                current_hunk = {
                    'old_start': old_line - (i - start_ctx),
                    'new_start': new_line - (i - start_ctx),
                    'lines': [],
                    'old_count': 0,
                    'new_count': 0
                }
                # Add context before
                for j in range(start_ctx, i):
                    op_ctx, line_ctx = diff_ops[j]
                    current_hunk['lines'].append(f" {line_ctx}")
                    current_hunk['old_count'] += 1
                    current_hunk['new_count'] += 1

            # Add the change
            if op == '+':
                current_hunk['lines'].append(f"+{line}")
                current_hunk['new_count'] += 1
            else:
                current_hunk['lines'].append(f"-{line}")
                current_hunk['old_count'] += 1

        else:
            # Context line
            if current_hunk is not None:
                # Check if we should close the hunk
                # Look ahead to see if there's another change within context range
                next_change = None
                for j in range(i + 1, min(i + context * 2 + 1, len(diff_ops))):
                    if diff_ops[j][0] != ' ':
                        next_change = j
                        break

                if next_change is not None and next_change - i <= context * 2:
                    # Continue the hunk
                    current_hunk['lines'].append(f" {line}")
                    current_hunk['old_count'] += 1
                    current_hunk['new_count'] += 1
                else:
                    # Add trailing context and close hunk
                    trailing_ctx = 0
                    for j in range(i, min(i + context, len(diff_ops))):
                        op_ctx, line_ctx = diff_ops[j]
                        if op_ctx == ' ':
                            current_hunk['lines'].append(f" {line_ctx}")
                            current_hunk['old_count'] += 1
                            current_hunk['new_count'] += 1
                            trailing_ctx += 1
                        else:
                            break

                    hunks.append(DiffHunk(
                        old_start=current_hunk['old_start'],
                        old_count=current_hunk['old_count'],
                        new_start=current_hunk['new_start'],
                        new_count=current_hunk['new_count'],
                        lines=current_hunk['lines']
                    ))
                    current_hunk = None
                    i += trailing_ctx - 1  # Skip the trailing context we just added

        # Update line counters
        if op in (' ', '-'):
            old_line += 1
        if op in (' ', '+'):
            new_line += 1

        i += 1

    # Close any remaining hunk
    if current_hunk is not None:
        hunks.append(DiffHunk(
            old_start=current_hunk['old_start'],
            old_count=current_hunk['old_count'],
            new_start=current_hunk['new_start'],
            new_count=current_hunk['new_count'],
            lines=current_hunk['lines']
        ))

    return hunks


def compute_diff_stat(old_lines: List[str], new_lines: List[str]) -> Tuple[int, int]:
    """Compute number of insertions and deletions."""
    diff_ops = lcs_diff(old_lines, new_lines)
    insertions = sum(1 for op, _ in diff_ops if op == '+')
    deletions = sum(1 for op, _ in diff_ops if op == '-')
    return insertions, deletions
